### ISSUE TITLE
[HOLD] [$16,000] [Performance] Loading the personalDetailsList takes a long time for some users

### ISSUE THREAD

**Issue by marcaaron** [Expensify/App#7942]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## What performance issue do we need to solve?

The query to get personalDetails can be quite slow when you have access to tons of user details like huge accounts do. We have optimized this query several times. The main problem here is that the app blocks on loading these details when they aren't critical to launching the app. 

## What is the impact of this on end-users?

TTFB for `Get&returnValueList=personalDetailsList` is pretty bad which can lead to very slow initializing of the app.

## List any benchmarks that show the severity of the issue

Log search:

David getting `personalDetailsList` - 5581ms 😱 (most I sampled today are closer to 1500ms but that is proof it can be QUITE slow)  - [Logs](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%226e45e9747b5a1392-LAX%22)+AND+timestamp:[2022-02-28T00:34:46.974Z+TO+2022-02-28T02:34:46.974Z]&index=logs_expensify-015599)
Marc getting `personalDetailsList` - 31 ms - [Logs](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id:(%226e4b33871d06f4dd-SJC%22)+AND+timestamp:[2022-02-28T15:59:04.255Z+TO+2022-02-28T17:59:04.255Z]&index=logs_expensify-015613)

## Proposed solution (if any)

We should load personal details "on demand", and cache them in some fashion. I think we should do everything we can to never block app load on an API call unless it is critical (like signing in).

The way we could rework all this could be something like:

1. Remove the current API call
2. Add a new non-blocking API call that only gets the personal details for reports in LHN. When the data comes in, it is stored in Onyx.
    1. If you are in focus mode, it will load all people from LHN first, but if you're in most recent mode, it would start with only the top 50 reports (assuming that things below the fold can be updated last).
3. Add a new non-blocking API call that gets all the personal details (what is currently done). When the data comes in, it is stored in Onyx (it's fine to delay this for later, like 30 seconds later or something)
4. The LHN will always default to showing the basic login (email address or phone number) and default avatar. Whenever personal details are updated in Onyx, then the LHN automatically re-renders to show the names and avatars
5. Update all personal details every 30 minutes

## List any benchmarks after implementing the changes to show impacts of the proposed solution (if any)

We can ask David to test this change to see if the app loads faster for him.

## Platform:
Where is this issue occurring?

- [x] Web
- [x] iOS
- [x] Android
- [x] Desktop App
- [x] Mobile Web

**Version Number:** All
**Logs:** N/A
**Notes/Photos/Videos:** N/A
**Expensify/Expensify Issue URL:** N/A

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mountiny (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by mountiny** [comment]
As far as I know, that is the known mega query that takes super long. I won't be able to look into this as I will be ooo most of the time in the coming weeks. I will ping in Slack and also create a duplicate in Expensify repo so this does not fall through cracks based on the process we would like to possibly follow in [future](https://expensify.slack.com/archives/C03TQ48KC/p1645637474144999?thread_ts=1645630309.249429&cid=C03TQ48KC).

**Comment by mountiny** [comment]
Internal issue create [here](https://github.com/Expensify/Expensify/issues/199319).

Asked in Slack [here](https://expensify.slack.com/archives/C03TQ48KC/p1646083971841459).

**Comment by flodnv** [comment]
The last optimization was made here: https://github.com/Expensify/Auth/pull/6111, more details can be found in [this slack channel](https://expensify.slack.com/archives/C02JWALB9PF/p1636396432397700). 

`5581ms` : that server was having known problems we're working on at that time, so this is not normal operation. 
`1500ms` : we can try finding ways to make this faster, it'll be challenging but certainly not impossible. What is our target @marcaaron ? 

>This is a backend issue that disproportionately affects users with large numbers of reports.

_How_ exactly does it affect users? Is it blocking app load or something like that? Are we trying to get all personal details of everyone "connected" to you? If yes, why?

**Comment by marcaaron** [comment]
Lots of great questions @flodnv. FWIW I'm just reporting anything that looks remotely suspect here.

It gets called very early in the app init

https://github.com/Expensify/App/blob/53581fef99e0df99f5c93a5fb5be0dbd89d3925b/src/libs/Navigation/AppNavigator/AuthScreens.js#L119-L119

And the data is used to populate the sidebar with names, avatars, etc.

>How exactly does it affect users? Is it blocking app load or something like that? 

Since this is not slow for me the test I performed (that anyone can also try) is to noop (or add a long `setTimeout`) to the function [here](https://github.com/Expensify/App/blob/00805b2b495fac25f53a42231dd3e5ae78138a2a/src/libs/actions/PersonalDetails.js#L123-L149) to see how the app looks in the time before it loads and then test against the production API to observe any bad effects. It just shows a white screen for a while...

>Are we trying to get all personal details of everyone "connected" to you? If yes, why?

This I don't know.

>What is our target @marcaaron ?

I'm not sure about this either. There might be other solutions here we're not considering yet and there's room for creativity. 

e.g. coming up with a more specific command that returns exactly what we need or splitting things up so that we get the entire personal details list late, but some much smaller subset of user personals early (basically, load what is critical first then load everything else later).

**Comment by flodnv** [comment]
Yes, I was thinking that we should load personal details "on demand", and cache them in some fashion. I think we should do everything we can to never block app load on an API call unless it is critical (like signing in) -- would you agree with that premise @marcaaron @iwiznia @tgolen @AndrewGable ? 

The way we could rework all this could be something like:
1. Get rid of this API call blocking the whole UI
2. On first fresh app load, load the personal details of all avatars we need to display in the LHN. Cache them in some way.
3. In the background, fetch all personal details of everyone (eg the API call removed in 1). Cache them in some way.
4. On the next app load, use the cached personal details -- they are already loaded, so they can't be blocking the app load.
5. When the app is loaded, go to (3)

Would something like this be doable?

Writing this out also makes me realize: if I update my avatar, when will you see the updated one? After you reload your app?

**Comment by marcaaron** [comment]
>On first fresh app load, load the personal details of all avatars we need to display in the LHN. Cache them in some way.

To clarify, this does happen already. But only after you are signed in (subsequent loads / refreshes of the site are fast). But we can still make the first sign in faster by loading the critical personal details quickly.

>if I update my avatar, when will you see the updated one? After you reload your app?

Correct, personal details do not update in realtime IIRC.

**Comment by flodnv** [comment]
>we can still make the first sign in faster by loading the critical personal details quickly.

Ok so can we easily get this done from an external contributor? I think this is a much better route than trying to optimize the query that we already super optimized -- I am not sure how much more juice we can get out of this exercise. When you'll have DMs and reports and policies with thousands of people, it is currently kind of impossible to make it near instant.

**Comment by marcaaron** [comment]
Yes of course, please feel free to update the labels, description, add design notes or anything else you think would help. I don't have any specific ideas for a solution, but sounds like I achieved my goal of drawing attention to the problem.

**Comment by tgolen** [comment]
The personal details list refreshes every 30 minutes on a timer, so that's
when you would get the updated avatar. I'm fine with everything else
mentioned in this issue.

**Comment by iwiznia** [comment]
I agree with the plan. To clarify, nothing should block the app from loading, so 2 would be non blocking, right? LHN would render with emails first and then when personal details are loaded into onyx, the component would update and render with the names, right?

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days.  eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by flodnv** [comment]
@marcaaron @tgolen @iwiznia I have updated the issue description with what we decided, please check it to make sure we all agree on what we want to do.

**Comment by marcaaron** [comment]
Yes, sounds great to me!

**Comment by tgolen** [comment]
I read through it and I'm a little confused on a couple of points (assume that I agree with all the rest that I don't comment on).

>  Cache them in some way

This is mentioned twice, but it just means they will be put into Onyx, right?

> On first fresh app load

How will we know that this is a "first fresh app load"? Does this just mean that the app has loaded and there is no data in Onyx? How will we know that?

> On the next app load

I think this means something more like "if the app is loaded and data already exists in Onyx"? How will we know that?

> When the app is loaded, go to (3)

This really has my head spinning and I am not sure how to resolve the confusion, but it's some weird loop where I don't know what state this belongs to (first free app load or next app load or some other app load)


I think I would simplify it like this:

1. Remove the current API call
2. Add a new non-blocking API call that only gets the personal details for reports in LHN. When the data comes in, it is stored in Onyx.
3. Add a new non-blocking API call that gets the personal details for reports NOT in the LHN. When the data comes in, it is stored in Onyx (it's fine to delay this for later, like 30 seconds later or something)
4. The LHN will always default to showing the basic login (email address or phone number) and default avatar. Whenever personal details are updated in Onyx, then the LHN automatically re-renders to show the names and avatars

**Note:** I am a bit concerned that this only improves performance if someone is in focus mode. If you are not in focus mode, then the LHN has all reports, and `1` would have ALL results while `3` has zero results.

**Comment by flodnv** [comment]
>This is mentioned twice, but it just means they will be put into Onyx, right?

I assume so, yes (if that makes sense to you) 

>How will we know that this is a "first fresh app load"? Does this just mean that the app has loaded and there is no data in Onyx? How will we know that?

Good question. Maybe we can just change it to "on app load" and it will still work just as good (and maybe in a simpler way)? 

Heh, yes, your updated proposal is great 👍  Editing the OP now. 

>If you are not in focus mode, then the LHN has all reports

Hmmm, not necessarily / not always. In "most recent" mode, it is not necessarily a guarantee that you have everyone's avatar displaying in the LHN. Even if you do (which I think is really unlikely), it will still improve performance of _startup_ since the API call will be non blocking.

**Comment by tgolen** [comment]
Maybe if you are in focus mode, it would load all people from LHN first, but if you're in most recent mode, it would start with only the top 10 reports or something (assuming that things below the fold can be updated last).

**Comment by quinthar** [comment]
FYI, I'm in focus mode.

**Comment by iwiznia** [comment]
I don't get why the new step 3 loads details of reports not in the LHN instead of all details. Especially, keeping in mind that we want all details even of people were you share no reports with (so their data shows when starting a new chat for example)

**Comment by tgolen** [comment]
My main goal was to be sure step 3 doesn't want to repeat the same data requested in step 2. Rather than doing a subset of everything and then everything, I think it should be requesting a subset, and then requesting everything else that is not in that subset.

**Comment by iwiznia** [comment]
> I think it should be requesting a subset, and then requesting everything else that is not in that subset.

I am fine with that, just note it's not what's written in the proposal. Also, maybe returning everything not in the subset is more complex or takes more time than just retrieving everything, so if it is, then let's just return everything.

**Comment by flodnv** [comment]
>Maybe if you are in focus mode, it would load all people from LHN first, but if you're in most recent mode, it would start with only the top 10 reports or something (assuming that things below the fold can be updated last).

Agreed, adding this.

>I am fine with that, just note it's not what's written in the proposal. Also, maybe returning everything not in the subset is more complex or takes more time than just retrieving everything, so if it is, then let's just return everything.

I agree returning everything is less error prone (and probably simpler) than the other option, updating this.

**Comment by adelekennedy** [comment]
@iwiznia @flodnv @tgolen Should this still be an external issue? Ready to open an upwork job if contributors should work on this!

**Comment by flodnv** [comment]
I think yes after @iwiznia & @tgolen give their 👍

**Comment by tgolen** [comment]
Yeah, I think it is OK to be external. There are no API changes that need to be done on the server for this proposal.

**Comment by iwiznia** [comment]
👍 from me, but don't we need a new api command or param to only return details of people in the LHN for step 2?

**Comment by flodnv** [comment]
I think we can use [`PersonalDetails_GetForEmails`](https://github.com/Expensify/App/blob/cb1925c5f011dfe555a4ac0d75ee24da58c17388/src/libs/API.js#L449)

**Comment by tgolen** [comment]
Yeah, I was assuming that the front-end would just be passing whatever emails it wants to get the details for.

**Comment by tgolen** [comment]
However, the thought of moving all that logic to Auth is really a good idea and I think that would be awesome!

**Comment by iwiznia** [comment]
I see, anyway, I am not sure which one is better than the other, I think I lean slightly to the frontend as it feels more correct to be handled there

**Comment by adelekennedy** [comment]
[Internal](https://www.upwork.com/ab/applicants/1509990319706157056/job-details)
[External](https://www.upwork.com/jobs/~015af2f14cc7404530)

**Comment by Santhosh-Sellavel** [comment]
@roryabraham Can we double this one it's been 10days?

cc: @adelekennedy

**Comment by roryabraham** [comment]
Yes, I think we can double this

**Comment by adelekennedy** [comment]
Just got the Slack ping to double, updating now!

**Comment by adelekennedy** [comment]
doubling again - waiting for proposals

**Comment by adelekennedy** [comment]
still no proposals - moving forward to double

**Comment by Santhosh-Sellavel** [comment]
Waiting for a proposal!

**Comment by adelekennedy** [comment]
Doubling time

**Comment by adelekennedy** [comment]
and it's time to double again 😅

**Comment by adelekennedy** [comment]
fine with me - adding a note for myself here that we are opting **not to double** over the next few weeks (potentially putting the issue on hold)

**Comment by iwiznia** [comment]
ok yes, let's put this on hold. Jason agreed [here](https://expensify.slack.com/archives/C02HWMSMZEC/p1653413805745199?thread_ts=1653338396.350779&cid=C02HWMSMZEC) too.

**Comment by adelekennedy** [comment]
deleting the doubling announcement

**Comment by roryabraham** [comment]
Still on HOLD for the [App Start / Reconnect API refactor doc](https://docs.google.com/document/d/1Mc2fvigSqPnMa6yv6FQk7_opRpjK9kg0e2Gzz9QTXWM/edit)

**Comment by adelekennedy** [comment]
still on hold

**Comment by roryabraham** [comment]
A lot has changed since this issue was created. It's not clear to me that this is still relevant in its current form, since we now have single `OpenApp` and `ReconnectApp` API endpoints that provide personalDetails data. Furthermore, that data does not need to be transformed by the front-end at all anymore since the API provides it in the format that the front-end needs.

The caching solution that @flodnv talked about should also already be in place – personal details are saved to Onyx, loaded from disk, and fetched in the background.

AFAICT, the only remaining pieces that we can do to speed things up/create a better UX is:

1. Speed up the `getJSONPersonalDetailsList` query, if possible 🤷
1. Prevent the UI from blocking when `personalDetails` are not fully loaded (most relevant on a fresh sign-in). In practice, this would probably mean implementing the skeleton UI for more components that are dependent upon `personalDetails` data to render correctly, most notably the LHN.

Does that sound correct to everyone? If so, I'd recommend that we actually close this and create new separate issues for each with fresh bounties. `#2` can be external, while `#1` would need to be internal.

**Comment by flodnv** [comment]
Thanks for the update, sounds great to me 👍 I think we should do `#2` and then evaluate if we need `#1` at all, as it will be much more challenging (has already been done a handful of times, so we're running out of options, but this is not to say that it's impossible)

**Comment by Santhosh-Sellavel** [comment]
@roryabraham Yeah good for me, If possible I would like to be C+ for the new issue which will be created.

**Comment by roryabraham** [comment]
> Prevent the UI from blocking when personalDetails are not fully loaded (most relevant on a fresh sign-in). In practice, this would probably mean implementing the skeleton UI for more components that are dependent upon personalDetails data to render correctly, most notably the LHN.

While this sounds good, I need to spend some time investigating to get crisp on the exact requirements here (it might actually end up being a pretty big change). Going to bump this to weekly so I get to it sooner rather than later.

**Comment by adelekennedy** [comment]
still on hold

**Comment by adelekennedy** [comment]
@roryabraham this is still on hold? Should we move it to monthly or is the weekly label helpful?

**Comment by adelekennedy** [comment]
Seems liek this is still on hold

**Comment by roryabraham** [comment]
@adelekennedy this is held on me listing out components that are dependent upon personalDetails to render and determining which ones we should use skeleton UI on (where personalDetails are not guaranteed to be provided and are used for rendering purposes).

Here's a list of components that are currently _using_ personal details, but I haven't qualified which are not guaranteed to have them at the time they render or which are actually rendering stuff taken from personalDetails:

- `ArchivedReportFooter`
- `ReportWelcomeText`
- `InitialSettingsPage`
- `IOUConfirmationList`
- `AdditionalDetailsStep`
- `RequestCallPage`
- `ReportActionCompose`
- `SidebarLinks`
- `IOUModal`
- `IOUPreview`
- `DateUtils`
- `DetailsPage`
- `NewChatPage`
- `ReportDetailsPage`
- `ReportParticipantsPage`
- `SearchPage`
- `HeaderView`
- `ReportActionItemCreated`
- `SidebarLinks`
- `IOUParticipantsPage`
- `IOUParticipantsSplit`
- `WorkspaceInvitePage`
- `WorkspaceMembersPage`

**Comment by roryabraham** [comment]
More work is needed, but ultimately the next step here is to create new issues to render the skeleton UI when we're waiting for personalDetails, and then close this out. No external job will be associated with this issue.

**Comment by roryabraham** [comment]
At this point, I'm not sure the continued relevance of this issue. Yes we can (and probably should) implement the skeleton UI in more places across the app, but I think the best next step towards that would come in the form of a full-throttle to nail down exactly where it's possible to see old-school loading spinners or just blank UI and replace that with skeleton UI stuff. Going to unassign myself at this point. Seems like a great project for a new Bug0 member looking to pick up a more comprehensive tracking-issue style project.

**Comment by puneetlath** [comment]
I'm going to close this. We have plenty of bugs to focus on and it seems there isn't agreement here. Feel free to reopen if you disagree.
